### ISSUE TITLE
Performance optimization of MOOSE framework

### ISSUE THREAD

**Issue by roystgnr** [idaholab/moose#10256]
## Rationale

MOOSE could be and should be faster.  I believe much of the work needed to make that happen can be done entirely at the libMesh level, but there will be cases (where MOOSE itself is being inefficient, or where future libMesh changes provide more efficient ways to do things by using new APIs) where optimization will require changes to MOOSE code as well.

## Description

Unless it's too broad, I'd like to make this a catch-all ticket for performance improvements in MOOSE.

Reproducing poor performance reliably, for MOOSE as a whole, is still an open question, currently being addressed in #9265.

At the moment I'm using the PrairieDog test case provided by @jessecarterMOOSE to look at AMR/C inefficiencies, using callgrind as the "last word" as to whether proposed optimizations are really improvements or not, but obviously callgrind results won't scale to larger and/or parallel problems.

## Impact

Ideally any changes made in support of this ticket will be entirely backwards-compatible, improving old API implementations or at worst adding new APIs while maintaining support for old ones.  If we ever encounter cases where that restriction seems too limiting then such cases probably deserve a separate ticket and discussion.

**Comment by permcody** [comment]
I'm ok with this ticket for the time being. We may close it after some period of time and several merges. We have other catch-all tickets but it's nice to close them eventually. Of course we can always reference closed tickets!

**Comment by mangerij** [comment]
Greatest ticket of all time.

**Comment by rwcarlsen** [comment]
We could really use one or two larger representative benchmark problems (~1-2 minutes in serial) to help us measure progress (or regress) toward this.  Suggestions?

**Comment by jessecarterMOOSE** [comment]
Yes, please. My App is not very representative of what the MOOSE framework can do. It doesn't even have Multiphysics! (Does that make it an OOSE app?) It's really just for testing mesh refinement.

Also, however "optimization" is done, let's also make sure we're testing at high cpu counts too.

**Comment by GiudGiud** [comment]
The MooseVariableData changes in #29697 got another 6-10% on sims that do basic FE and dont involve too many closures which I believe is what we are targetting here
